### ISSUE TITLE
CMake config doesn't obey -DLIB_SUFFIX

### ISSUE THREAD

**Issue by Bob131** [keystone-engine/keystone#213]
Hey! I was just writing up a spec file for keystone, and I noticed it was installing to `$PREFIX/lib` instead of `lib64`. I'm not really familiar with CMake, so all I can really say is that the Fedora cmake macro uses `-DLIB_SUFFIX` to indicate an arch-specific directory. From a cursory Google, it seems as though it's more a _de facto_ standard than anything official, but nonetheless it'd be great if you could update the config!

To save you a minute or two, this is the form the flag takes (excerpted from the macro file for clarity):
```
#
# Macros for cmake
#
%_cmake_lib_suffix64 -DLIB_SUFFIX=64
```

**Comment by aquynh** [comment]
can you make a pull request for this issue?

**Comment by Bob131** [comment]
Sorry, not really. I really know nothing about cmake. I had a poke 
around to see if the fix was obvious, but I have no idea

**Comment by aquynh** [comment]
so you had to put the below line somewhere in the source? exactly where did you put that?

    %_cmake_lib_suffix64 -DLIB_SUFFIX=64

**Comment by Bob131** [comment]
No, that line is automatically inserted into a spec file as part of the 
cmake set of macros. It's not really related though, I just mentioned 
it for some context. Specifically what this bug is about is that 
running cmake with LIB_SUFFIX set should result in makefiles that 
install into the proper directory, ie `cmake -DLIB_SUFFIX=64 .. && make 
install` should install libs into `$PREFIX/lib64`. That's all. Sorry 
for the confusion.

**Comment by aquynh** [comment]
i see. so this should be documented in docs/COMPILE-NIX.md.

is this for Fedora/Redhat?

**Comment by Bob131** [comment]
Yes, indeed it is. I have no idea how widespread the LIB_SUFFIX option 
is, but I imagine other distros use something similar.

**Comment by aquynh** [comment]
exactly how did you pass LIB_SUFFIX to cmake? i passed -DLIB_SUFFIX=64 to cmake 3.6 & 3.5.2, and both say:

```
CMake Warning:
  Manually-specified variables were not used by the project:

    LIB_SUFFIX
```

**Comment by aquynh** [comment]
to clarify, i tried above on Fedora 24, 64bit

**Comment by aquynh** [comment]
latest Github code now supports `lib64` option for `make-share.sh` & `make-lib.sh`, so you can simply do this on Fedora:

```
$ mkdir build
$ cd build
$ ../make-share.sh lib64
$ sudo make install
```

finally, you will have `libkeystone.so.*` installed into `/usr/local/lib64`. 

let me know if this fixes your issue.

**Comment by Bob131** [comment]
Well, mostly. It does, strictly speaking, but it isn't exactly obvious for end users wanting to invoke CMake themselves that `-DLLVM_LIBDIR_SUFFIX` is the flag that they want as opposed to just `-DLIBDIR_SUFFIX`.

I had checked out a new branch to try and make a workable PR, but I just can't get it to cooperate. The other thing I was thinking of doing was just sed-replacing `LLVM_LIBDIR_SUFFIX` with `LIBDIR_SUFFIX`, but I'm not sure if that's what you want.

You can probably close this if you don't care, your new commit is certainly workable. Thanks for your efforts!

**Comment by aquynh** [comment]
but i want to make it work, not "almost work".

can you tell me what is that LIBDIR_SUFFIX? that is certainly not supported by cmake, and not by LLVM either. the official option is LLVM_LIBDIR_SUFFIX, and that is why i made that change.

so exactly what is wrong now?

**Comment by Bob131** [comment]
The option for LLVM is `LLVM_LIBDIR_SUFFIX`; since your top-level CMakeLists.txt pulls in LLVM's CMake config, Keystone has that option too. However, the standard flag for everyone else is `LIBDIR_SUFFIX`. They're the same, ie both are not implemented by CMake itself but by a project's CMakeLists.txt. They have the same function too, the only difference is the name.

Right now, a tool or a human may use `-DLIBDIR_SUFFIX` expecting that to be the option required they need; that's how it is in a lot of other projects. The variable in the Keystone CMakeLists.txt is called `LLVM_LIBDIR_SUFFIX` because it has inherited it from the LLVM CMakeLists.txt. This might be appropriate if, eg, Keystone was an LLVM fork. But as Keystone simply _relies_ on LLVM, naming the option this way doesn't really make any sense.

At this point it's just a matter of providing users with the option names they expect; it's one less hurdle in the process of getting some awesome software :)

**Comment by aquynh** [comment]
i see. feel free to send a PR to add that option if you really want that, thanks.

**Comment by volkertb** [comment]
By the way, the same issue of lib64 not (or no longer) being a recognized parameter also ocrrus when building the static libraries by running `../make-lib.sh lib64`. I get exactly the same output as when trying to build the shared libraries with the lib64 parameter as described in my previous comment.

**Comment by volkertb** [comment]
Well, kstool seems to work fine on the command line after building, so omitting `lib64` during building on Fedora 64-bit systems seems to work fine.
